### PR TITLE
OAuth 2.0 docs

### PR DESCRIPTION
**Pull request by jasonblais**
We might want to have additional documentation on
 - more in-depth explanation of the OAuth 2.0 authorization flow
 - a concrete example of setting up a third-party app with Mattermost

If so, this would be a ticket for a dev to work on (at least the first draft)

### CODE REVIEW

File: source/developer/oauth2-applications.rst [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+OAuth 2.0 is a protocol that allows Mattermost to authorize API requests from external applications. The authorization lets 
+
+ - users with an account on a Mattermost server to sign in to third-party applications. You can find a `sample OAuth2 Client Application for Mattermost here <https://github.com/enahum/mattermost-oauth2-client-sample>`_ to test the functionality.
+
+ - a Mattermost server to authenticate requests to a third-party API. One popular application is Zapier integration which allows you to integrate more than 500 applications with Mattermost through OAuth 2.0. See our `Zapier documentation <https://docs.mattermost.com/integrations/zapier.html>`_ to learn more.
+
+This documentation includes the following details:
+
+ - `Register your application in Mattermost <https://docs.mattermost.com/developer/oauth-2-0-applications#register-your-application-in-mattermost>`_
+ - `Permissions to your application  <https://docs.mattermost.com/developer/oauth-2-0-applications#permissions-to-your-application>`_
```
Comment (esethna) [inline source/developer/oauth2-applications.rst]: Is this supposed to be "Managing" or "Granting" permissions to your application.

File: source/developer/oauth2-applications.rst [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+3 - Set **Is Trusted**: When set to **Yes**, your application is considered trusted by Mattermost. This means Mattermost doesn't require users to accept authorization when signing to third-party applications. When set to **No**, users will be provided with the following page to accept or deny authorization when authenticating for the first time.
+
+.. image:: ../images/oauth2_authorization_screen.PNG
+
+4 - Specify the **Display Name**: Enter a name for your application made of up to 64 characters. This is the name users will see when granting access to the application, when viewing a list of authorized applications in **Account Settings > Security > OAuth 2.0 Applications** and when viewing a list of OAuth 2.0 applications in **Main Menu** > **Integrations**.
+
+5 - Add **Description**: This is a short description of your application that users will see when viewing a list of authorized applications in **Account Settings > Security > OAuth 2.0 Applications**.
+
+6 - Specify the **Homepage**: This is the homepage of the OAuth 2.0 application. The URL must be a valid URL and start with ``http://`` or ``https://`` depending on your server configuration.
```
Comment (esethna) [inline source/developer/oauth2-applications.rst]: How is the homepage used?

File: source/developer/oauth2-applications.rst [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+9 - Click **Save** to create the application. 
+
+.. image:: ../images/oauth2_app_screen.png
+
+10 - You will be provided with a **Client ID**, **Client Secret** and the authorized redirect URLs. Save these values and use them in your application to connect it to Mattermost.
+
+.. image:: ../images/oauth2_confirmation_screen.PNG
+
+Note: **Client Secret** can be regenerated by the application creator or System Administrator, preventing new users from accessing the OAuth 2.0 application until the client secret has been updated in your app configuration.
```
Comment (esethna) [inline source/developer/oauth2-applications.rst]: Maybe note that the old secrets are not deauthorized

**Comment by esethna** [comment]
+1 other than the minor notes above. Nice doc!

File: source/developer/oauth2-applications.rst [hunk 4]
```diff
@@ -0,0 +1,80 @@
+OAuth 2.0 Applications
+===============
+
+OAuth 2.0 is a protocol that allows Mattermost to authorize API requests from external applications. The authorization lets 
+
+ - users with an account on a Mattermost server to sign in to third-party applications. You can find a `sample OAuth2 Client Application for Mattermost here <https://github.com/enahum/mattermost-oauth2-client-sample>`_ to test the functionality.
+
+ - a Mattermost server to authenticate requests to a third-party API. One popular application is Zapier integration which allows you to integrate more than 500 applications with Mattermost through OAuth 2.0. See our `Zapier documentation <https://docs.mattermost.com/integrations/zapier.html>`_ to learn more.
```
Comment (enahum) [inline source/developer/oauth2-applications.rst]: in the blog post you mention more than 700 applications.

File: source/developer/oauth2-applications.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+5 - Add **Description**: This is a short description of your application that users will see when viewing a list of authorized applications in **Account Settings > Security > OAuth 2.0 Applications**.
+
+6 - Specify the **Homepage**: This is the homepage of the OAuth 2.0 application. The URL must be a valid URL and start with ``http://`` or ``https://`` depending on your server configuration.
+
+7 - (Optional) Add **Icon URL**: The image users will see when viewing a list of authorized applications in **Account Settings > Security > OAuth 2.0 Applications** and when viewing a list of OAuth 2.0 applications in **Main Menu** > **Integrations**. Must be a valid URL and start with ``http://`` or ``https://``.
+
+8 - Add **Callback URLs**: This are the URL(s) to which Mattermost will redirect users after accepting or denying authorization of your application, and which will handle authorization codes or access tokens. Each URL must be on a separate line and start with ``http://`` or ``https://``.
```
Comment (enahum) [inline source/developer/oauth2-applications.rst]: Additional note: the Callbacks URL(s) are the only ones that can be used to authorize and app. Mattermost will only redirect to the **one URL** used when authorizing the app

File: source/developer/oauth2-applications.rst (hunk 3 above)
Comment (enahum) [inline source/developer/oauth2-applications.rst]: actually the tokens created with an old secret are still valid. If you regenerate the secret new authorizations will fail, also the request for a refresh token will fail unless the secret is updated in the OAuth2 app

File: source/developer/oauth2-applications.rst [hunk 6]
```diff
@@ -0,0 +1,80 @@
+OAuth 2.0 Applications
+===============
+
+OAuth 2.0 is a protocol that allows Mattermost to authorize API requests from external applications. The authorization lets 
+
+ - users with an account on a Mattermost server to sign in to third-party applications. You can find a `sample OAuth2 Client Application for Mattermost here <https://github.com/enahum/mattermost-oauth2-client-sample>`_ to test the functionality.
```
Comment (lfbrock) [inline source/developer/oauth2-applications.rst]: "Mattermost server to sign in" should have "to" removed

File: source/developer/oauth2-applications.rst [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+ - a Mattermost server to authenticate requests to a third-party API. One popular application is Zapier integration which allows you to integrate more than 700 applications with Mattermost through OAuth 2.0. See our `Zapier documentation <https://docs.mattermost.com/integrations/zapier.html>`_ to learn more.
```
Comment (lfbrock) [inline source/developer/oauth2-applications.rst]: remove "to"

File: source/developer/oauth2-applications.rst [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+.. image:: ../images/oauth2_authorization_screen.PNG
+
+4 - Specify the **Display Name**: Enter a name for your application made of up to 64 characters. This is the name users will see when granting access to the application, when viewing a list of authorized applications in **Account Settings > Security > OAuth 2.0 Applications** and when viewing a list of OAuth 2.0 applications in **Main Menu** > **Integrations**.
+
+5 - Add **Description**: This is a short description of your application that users will see when viewing a list of authorized applications in **Account Settings > Security > OAuth 2.0 Applications**.
+
+6 - Specify the **Homepage**: This is the homepage of the OAuth 2.0 application and let's users visit the app page to learn more what it does. The URL must be a valid URL and start with ``http://`` or ``https://`` depending on your server configuration.
```
Comment (lfbrock) [inline source/developer/oauth2-applications.rst]: "let's" > lets

**Comment by lfbrock** [comment]
+1 other than minor grammar comments

**Comment by jasonblais** [comment]
Thanks!

**Pull request merged**